test/phoenix_live_view/async_test.exs

```diff
@@ -0,0 +1,87 @@
+defmodule Phoenix.LiveView.AsyncTest do
+  # run with async: false to prevent other messages from being captured
+  use ExUnit.Case, async: false
+
+  import ExUnit.CaptureIO
+
+  describe "async operations" do
+    for fun <- [:assign_async, :start_async] do
+      test "warns when passing socket to #{fun} function" do
+        warnings =
+          capture_io(:stderr, fn ->
+            Code.eval_string("""
+            defmodule AssignAsyncSocket do
+              use Phoenix.LiveView
+
+              def mount(_params, _session, socket) do
+                {:ok, #{unquote(fun)}(socket, :foo, fn ->
+                  do_something(socket.assigns)
+                end)}
+              end
+
+              defp do_something(_socket), do: :ok
+            end
```

Comment:
You can probably use `eval_quoted`. The `if false` is because we want the code expanded but not executed. :)

```suggestion
            require Phoenix.LiveView

            if false do
              #{unquote(fun)}(socket, :foo, fn ->
                do_something(socket.assigns)
              end)
            end
```

Reply:
No luck with eval_quoted.

```elixir
ExUnit.start()

defmodule MyTest do
  use ExUnit.Case

  import ExUnit.CaptureIO

  for fun <- [:assign_async, :start_async] do
    test "warns when passing socket to #{fun} function" do
      warnings =
        capture_io(:stderr, fn ->
          Code.eval_quoted(
            quote bind_quoted: [fun: unquote(fun)] do
              require Phoenix.LiveView

              if false do
                fun(socket, :foo, fn ->
                  do_something(socket.assigns)
                end)
              end
            end
          )
        end)

      assert warnings =~
               "you are accessing the LiveView Socket inside a function given to #{unquote(fun)}"
    end
  end
end
```

```
mix run test.exs


  1) test warns when passing socket to start_async function (MyTest)
     test.exs:9
     ** (CompileError) nofile: cannot compile file (errors have been logged)
     stacktrace:
       (elixir 1.16.1) src/elixir_clauses.erl:45: :elixir_clauses.clause/6
       (elixir 1.16.1) src/elixir_clauses.erl:347: anonymous fn/7 in :elixir_clauses.expand_clauses_origin/6
       (stdlib 5.2) lists.erl:1706: :lists.mapfoldl_1/3
       (stdlib 5.2) lists.erl:1707: :lists.mapfoldl_1/3
       (elixir 1.16.1) src/elixir_clauses.erl:351: :elixir_clauses.expand_clauses_origin/6
       (stdlib 5.2) lists.erl:1706: :lists.mapfoldl_1/3
       (elixir 1.16.1) src/elixir_clauses.erl:98: :elixir_clauses.case/4
       (elixir 1.16.1) src/elixir_expand.erl:692: :elixir_expand.expand_case/5
       (elixir 1.16.1) expanding macro: Kernel.if/2
       nofile:1: (file)



  2) test warns when passing socket to assign_async function (MyTest)
     test.exs:9
     ** (CompileError) nofile: cannot compile file (errors have been logged)
     stacktrace:
       (elixir 1.16.1) src/elixir_clauses.erl:45: :elixir_clauses.clause/6
       (elixir 1.16.1) src/elixir_clauses.erl:347: anonymous fn/7 in :elixir_clauses.expand_clauses_origin/6
       (stdlib 5.2) lists.erl:1706: :lists.mapfoldl_1/3
       (stdlib 5.2) lists.erl:1707: :lists.mapfoldl_1/3
       (elixir 1.16.1) src/elixir_clauses.erl:351: :elixir_clauses.expand_clauses_origin/6
       (stdlib 5.2) lists.erl:1706: :lists.mapfoldl_1/3
       (elixir 1.16.1) src/elixir_clauses.erl:98: :elixir_clauses.case/4
       (elixir 1.16.1) src/elixir_expand.erl:692: :elixir_expand.expand_case/5
       (elixir 1.16.1) expanding macro: Kernel.if/2
       nofile:1: (file)


Finished in 0.01 seconds (0.01s on load, 0.00s async, 0.00s sync)
2 tests, 2 failures

Randomized with seed 369048
```

“errors have been logged”

Reply:
What about:

```
          Code.eval_quoted(
            quote do
              require Phoenix.LiveView

              if false do
                unquote(fun)(socket, :foo, fn ->
                  do_something(socket.assigns)
                end)
              end
            end
```

Reply:
Actually, sorry:

```
        Code.eval_quoted(
            quote do
              require Phoenix.LiveView

              if false do
                Phoenix.LiveView.unquote(fun)(socket, :foo, fn ->
                  do_something(socket.assigns)
                end)
              end
            end
```

Reply:
Nope. No luck with that either. But I know why I get no compiler errors... It's inside the capture_io.

Removing it yields

```
error: undefined variable "socket" (context MyTest)
└─ nofile:1
```

which kind of makes sense.

Besides that I think the defmodule code is actually alright. Might even be better to understand.

Reply:
I added some eval_quoted variants as well. Let me know which ones you like better.

Reply:
https://github.com/phoenixframework/phoenix_live_view/pull/3161/commits/c80b583c161ba23e246e972566cb7d6e42b253e6